Sprite render rework and palette lookup refactor

Holladiewal opened the pull request:
Reworks how sprite rendering is done to  be closer to the real hardware, but this is not exact replication.

Sprites are now precached at the  ~~beginning of the scanline (pixel 0)~~ end of the previous scanline and rendering the pixel is now reduced to asking the Sprite if it would produce a non-transparent pixel (sprites with higher priority will still block any other sprite from rendering on that pixel, even if they are hidden behind the background)

The palette lookup now uses an array created upon PPU init, instead of creating and destroying RGB objects twice every pixel.

Performance looks to be better now, and some rudimentary profiling shows a significant reduction of time spent in `render_sprite_pixel` (`render_sprite_pixel` is now faster than `render_bg_pixel` which was not true before!)

We also now set and clear the sprite overflow flag.

~~Bonus content: Fixed the CMakeSource.txt to compile joypad and debugger, so CMake builds are working again.~~

Holladiewal committed:
Main: Move PPU variable to outside of loop

There is no need to continually create and destroy references

Holladiewal committed:
PPU: Refactor Sprite evaluation

We evaluate  sprites a lot closer to how the NES would do it, but without the memory constraints
This improves by quite a bit and  makes us now held back by  background tile rendering

Holladiewal committed:
PPU: make color palette be global and not recreate RGB instances everytime we want to use a color

jntrnr commented:
You may need to merge from the latest main. Some of those cmake changes landed recently

jntrnr commented:
I just merged another PR that came in before this one. You may need to rebase some of your changes off of it.

Optimoos commented:
@Holladiewal  I'm running your changes in my branch already, but I'm bad with git so I just copy/pasted. I'm pretty sure the only thing I had to do was adjust to horizontal/vertical_pixel being member variables.

Holladiewal committed:
Adapt changes to incorporate previously merged PR

Nothing fancy to see here, please move on

Holladiewal committed:
PPU: Implement sprite overflow

This is the easy way, without proper emulation of the buggy evaluation
logic in the PPU.
